### PR TITLE
ENH: Add a status monitor to trace the running stage of subtask

### LINKED ISSUE #489: ENH: Add a check whether scheduling is stuck
Currently, tasks are scheduled based on the memory and computing resources of the nodes. Sometimes, the estimated resources may not meet the requirements of any worker. In this scenario, the scheduling service could get stuck, causing no tasks to execute and no responses. To handle this, a timeout mechanism is needed. If the scheduler remains stuck beyond a threshold, an error should be rasied directly.

### PR DESCRIPTION
**Pull request by Flying-Tom**
## What do these changes do?

Add a status monitor to trace the running stage of subtask, which will be used to terminate stale tasks, e.g. when requesting mem quota

## Related issue number

Fixes #489

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass

### CODE REVIEW

File: python/xorbits/_mars/services/scheduling/worker/execution.py [hunk 1]
```diff
@@ -44,6 +45,48 @@
...
+
+        if kw:  # pragma: no cover
+            pass
+
+    async def __post_create__(self):
+        pass
+
+    async def __pre_destroy__(self):
+        pass
+
```
Comment (aresnow1) [inline python/xorbits/_mars/services/scheduling/worker/execution.py:62]: Do not need to define them.

**Review by aresnow1 — COMMENTED** [review]
Seems come code not covered.

File: python/xorbits/_mars/services/scheduling/worker/execution.py [hunk 2]
```diff
@@ -44,6 +45,48 @@
 
 logger = logging.getLogger(__name__)
 
+
+class StatusMonitorActor(mo.Actor):
+    def __init__(self, **kw):
+        super().__init__()
+        self._records = dict()
+
+        if kw:  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (Flying-Tom) [reply python/xorbits/_mars/services/scheduling/worker/execution.py:54]: To remove these placeholder-like codes as prior

File: python/xorbits/_mars/services/scheduling/worker/execution.py [hunk 3]
```diff
@@ -44,6 +45,48 @@
...
+            pass
+
+    async def __post_create__(self):
+        pass
+
+    async def __pre_destroy__(self):
+        pass
+
+    def report_status(self, keys: tuple[str, str], status: str):
```
[the remark this answers is not in the document]
Comment (Flying-Tom) [reply python/xorbits/_mars/services/scheduling/worker/execution.py:63]: Considering this as a simple status monitor which don't need complete stage info to handle #489 , Enum class needs some accurate status primitive which I can't define

File: python/xorbits/_mars/services/scheduling/worker/execution.py [hunk 4]
```diff
@@ -44,6 +45,48 @@
...
+                and v["history"][-1][1] == status
+            ):
+                stale_tasks_keys.append(k)
+
+        return stale_tasks_keys
+
+    async def get_records(self):
+        return self._records
+
+    def print_records(self):
```
[the remark this answers is not in the document]
Comment (Flying-Tom) [reply python/xorbits/_mars/services/scheduling/worker/execution.py:85]: These two methods work for testing in this pre-PR and will be removed in a formal version

File: python/xorbits/_mars/services/scheduling/worker/quota.py [hunk 5]
```diff
@@ -332,14 +338,37 @@ async def update_mem_stats(self):
...
-            # memory usage reduced: try reallocate existing requests
+            # memory usage reduced: try to reallocate existing requests
             await self._process_requests()
         self._last_memory_available = cur_mem_available
         self._report_quota_info()
         self._stat_refresh_task = self.ref().update_mem_stats.tell_delay(
             delay=self._refresh_time
         )
 
+        STALE_TIMEOUT = 5
```
[the remark this answers is not in the document]
Comment (Flying-Tom) [reply python/xorbits/_mars/services/scheduling/worker/quota.py:349]: Does this mean that the handling of 'stuck' should be done in following PRs, not in this one?

File: python/xorbits/_mars/services/scheduling/worker/execution.py (hunk 1 above)
Comment (Flying-Tom) [reply python/xorbits/_mars/services/scheduling/worker/execution.py:62]: These useless codes are used as placeholders for future functionality and will not appear in the last commit.

File: python/xorbits/_mars/services/scheduling/worker/quota.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aresnow1) [reply python/xorbits/_mars/services/scheduling/worker/quota.py]: Yep.

File: python/xorbits/_mars/services/scheduling/worker/execution.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aresnow1) [reply python/xorbits/_mars/services/scheduling/worker/execution.py]: enum avoids typos and allows knowing all the possible states. What exactly do you mean by accurate status?

File: python/xorbits/_mars/services/scheduling/worker/quota.py [hunk 6]
```diff
@@ -415,22 +421,14 @@ async def __post_create__(self):
...
-                )
-            else:
-                self._band_quota_refs[band] = await mo.create_actor(
-                    QuotaActor,
-                    band,
-                    quota_size=band_config["quota_size"],
-                    uid=QuotaActor.gen_uid(band[1]),
-                    address=self.address,
-                )
+            actor_cls = MemQuotaActor if hard_limit else QuotaActor
```
Comment (aresnow1) [inline python/xorbits/_mars/services/scheduling/worker/quota.py:424]: These changes should be reverted.

**Comment by aresnow1** [comment]
It's better to remove records if subtask is finished, they will take too much memory.

File: python/xorbits/_mars/services/subtask/core.py [hunk 7]
```diff
@@ -36,6 +36,16 @@
...
 
 
+class SubtaskStage(Enum):
+    PREPARE = 0
+    QUOTA = 1
+    SLOT = 2
+    EXECUTE = 3
+    FINALLY = 4
+    IMMEDIATELY = 5
+    FINISHED = 6
```
Comment (aresnow1) [inline python/xorbits/_mars/services/subtask/core.py:46]:
```suggestion
class SubtaskStage(Enum):
    PREPARE_DATA = 0
    REQUEST_QUOTA = 1
    ACQUIRE_SLOT = 2
    EXECUTE = 3
    RELEASE_SLOT = 4
    FINISH = 5
```

Comment (aresnow1) [reply python/xorbits/_mars/services/subtask/core.py:46]: IMMEDIATELY is not used.

**Review by aresnow1 — COMMENTED** [review]
The title of pull request need to be changed.

**Comment by Flying-Tom** [comment]
> The title of pull request need to be changed.

Title has been changed

**Review by aresnow1 — APPROVED** [review]
LGTM

**Pull request merged**